lib/Sema/CSGen.cpp

```diff
@@ -3102,6 +3102,18 @@ namespace {
       llvm_unreachable("found KeyPathDotExpr in CSGen");
     }
 
+    Type visitPoundAssertExpr(PoundAssertExpr *PAE) {
+      // Constrain the condition argument to be Bool.
+      auto boolType = CS.getASTContext().getBoolDecl()->getDeclaredType();
+      CS.addConstraint(
+          ConstraintKind::ArgumentConversion, CS.getType(PAE->getCondition()),
+          boolType,
+          CS.getConstraintLocator(PAE, ConstraintLocator::ApplyArgument));
+
+      // #assert(...) has type Void.
+      return CS.getASTContext().TheEmptyTupleType;
+    }
```

Comment:
Can we use `TypeChecker::typeCheckStmtCondition` to type check the pound condition just like for other statements if we module #assert as a statement instead of an expression like @rintaro suggested above? Generating constraint conditions seems to me an overkill for this case. cc: @xedin

Reply:
I agree, it seems like we are to go statement route, it shouldn't require any type-checker changes, but should use existing `tyeCheckStmtCondition` mechanism.